src/test/java/com/sevendragons/practice/leetcode/medium/courseschedule/Solution.java

```diff
@@ -0,0 +1,61 @@
+package com.sevendragons.practice.leetcode.medium.coursesschedule;
+
+import java.util.*;
+
+public class Solution {
+
+    //numCourses : id of course (ex: 3)
+    //prerequesite: depencencies of course (ex: [[3,2],[2,1],[1,2], [3,1]])
+    public boolean canFinish(int numberOfCourse, int[][] prerequisites) {
+
+        // Init graph
+        Map<Integer, Set<Integer>> prerequisitePerCourse = new HashMap<>();
+        for (int[] pair : prerequisites) {
+            if (prerequisitePerCourse.get(pair[0]) == null) {
+                prerequisitePerCourse.put(pair[0], new HashSet<Integer>());
+            }
+            prerequisitePerCourse.get(pair[0]).add(pair[1]);
+        }
+
+        // Detect cycle
+        return !isCyclic(numberOfCourse, prerequisitePerCourse);
+    }
+
+    private static boolean isCyclic(int numberOfCourse, Map<Integer, Set<Integer>> prerequisitePerCourse) {
+
+        for (int course = 0; course < numberOfCourse; course++) {
+            Set<Integer> targetCourses = new HashSet<Integer>();
+            targetCourses.add(course);
+            if (isCourseContainedInPrerequisites(prerequisitePerCourse, targetCourses, course) == true) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean isCourseContainedInPrerequisites(Map<Integer, Set<Integer>> prerequisitePerCourse, Set<Integer> targetCourses, int prerequisiteToCheck) {
+
+        for (int course : targetCourses) {
+            if (prerequisitePerCourse.get(prerequisiteToCheck) != null && prerequisitePerCourse.get(prerequisiteToCheck).contains(course)) {
```

Comment:
On this line line, you perform a hash map lookup for the same key twice. It would be better to do it only once.